upup/pkg/fi/cloudup/apply_cluster.go

```diff
@@ -675,12 +585,89 @@ func (c *ApplyClusterCmd) Run(ctx context.Context) error {
 		})
 
 	default:
-		return fmt.Errorf("unknown cloudprovider %q", cluster.Spec.CloudProvider)
+		return nil, fmt.Errorf("unknown cloudprovider %q", cluster.Spec.CloudProvider)
 	}
 
 	l.TemplateFunctions["Masters"] = tf.MasterInstanceGroups
 
 	err = tf.AddTo(l.TemplateFunctions, secretStore)
+	if err != nil {
+		return nil, err
+	}
+
+	return l, nil
+}
+
+func (c *ApplyClusterCmd) Run(ctx context.Context) error {
+	assetBuilder := assets.NewAssetBuilder(c.Cluster, string(c.Phase))
+
+	// Note that upgradeSpecs writes back to c.Cluster
+	if err := c.upgradeSpecs(assetBuilder); err != nil {
+		return err
+	}
+
+	cluster := c.Cluster
+
+	cloud, err := BuildCloud(cluster)
+	if err != nil {
+		return err
+	}
+
+	keyStore, err := c.Clientset.KeyStore(cluster)
+	if err != nil {
+		return err
+	}
+
+	sshCredentialStore, err := c.Clientset.SSHCredentialStore(cluster)
+	if err != nil {
+		return err
+	}
+
+	secretStore, err := c.Clientset.SecretStore(cluster)
+	if err != nil {
+		return err
+	}
+
+	l, err := c.buildLoader(ctx, cloud, keyStore, secretStore, sshCredentialStore, assetBuilder)
+	if err != nil {
+		return err
+	}
+
+	configBase, err := vfs.Context.BuildVfsPath(cluster.Spec.ConfigBase)
+	if err != nil {
+		return fmt.Errorf("error parsing config base %q: %v", cluster.Spec.ConfigBase, err)
+	}
+
+	if !c.AllowKopsDowngrade {
```

Comment:
In the existing code this and `configBase` came before `BuildCloud()`, etc. Why move it later?